### PR TITLE
Move JIT tests to gtest

### PR DESCRIPTION
**Pull request by goldsborough**
In our #better-engineering quest of removing all uses of catch in favor of gtest, this PR ports JIT tests to gtest. After #11846 lands, we will be able to delete catch. 

I don't claim to use/write these tests much (though I wrote the custom operator tests) so please do scrutinize whether you will want to write tests in the way I propose. Basically:

1. One function declaration per "test case" in test/cpp/jit/test.h
2. One definition in test/cpp/jit/test.cpp
3. If you want to be able to run it in Python, add it to `runJitTests()` which is called from Python tests
4. If you want to be able to run it in C++, add a `JIT_TEST` line in test/cpp/jit/gtest.cpp

Notice also I was able to share support code between C++ frontend and JIT tests, which is healthy.

@ezyang @apaszke @zdevito

### CODE REVIEW

**Comment by goldsborough** [comment]
@ezyang The tests from `tests.cpp` and the `runJitTests()`  function get linked into PyTorch, and exposed to Python. The assertions are changed to use `AT_ASSERT` instead of Gtest asserts. This structure was already how it was, I didn't really change anything here except that "catch-mode" changed to "gtest-mode"

Gtest doesn't get linked into Pytorch (not weird), it gets linked into the test binaries

File: test/cpp/jit/gtest.cpp [hunk 1]
```diff
@@ -0,0 +1,42 @@
+#include <gtest/gtest.h>
+
+#include <test/cpp/jit/tests.h>
+
+using namespace torch;
+using namespace torch::jit;
+
+#if defined(USE_GTEST)
```
[the remark this answers is not in the document]
Comment (goldsborough) [reply test/cpp/jit/gtest.cpp]: it shouldn't be in the file at all actually, I forgot to remove it

**Review by ezyang — APPROVED** [review]
Sure, why not.

It's pretty weird we need the Python-enabled version, but c'est la vie, I suppose.

**Comment by goldsborough** [comment]
@peterjc123 @yf225 do you know what I have to do here for Windows? I added the `TORCH_API` macro but it doesn't seem to apply ...

**Pull request closed without merging**